File: Bot.cs
// Bot.cs
using System;
using System.Threading.Tasks;
using Telegram.Bot;
using Telegram.Bot.Polling;
using Telegram.Bot.Types;
using Telegram.Bot.Types.Enums;
using Telegram.Bot.Types.ReplyMarkups;

class Bot
{
    private static readonly TelegramBotClient BotClient;
    private static readonly Stats Stats = new Stats();

    static Bot()
    {
        string token = Environment.GetEnvironmentVariable("BOT_TOKEN");
        if (string.IsNullOrEmpty(token))
            throw new Exception("BOT_TOKEN environment variable not set");
        BotClient = new TelegramBotClient(token);
    }

    class Stats
    {
        public int Messages { get; set; }
        public int Commands { get; set; }
    }

    static async Task Main()
    {
        using var cts = new CancellationTokenSource();
        var receiverOptions = new ReceiverOptions
        {
            AllowedUpdates = new[] { UpdateType.Message, UpdateType.CallbackQuery }
        };

        BotClient.StartReceiving(
            HandleUpdateAsync,
            HandlePollingErrorAsync,
            receiverOptions,
            cancellationToken: cts.Token
        );

        Console.WriteLine("Bot started. Press Ctrl+C to stop.");
        await Task.Delay(-1);
    }

    static async Task HandleUpdateAsync(ITelegramBotClient client, Update update, CancellationToken cancellationToken)
    {
        if (update.Message is Message msg)
        {
            long chatId = msg.Chat.Id;
            string? text = msg.Text;

            if (text == null) return;

            if (text.StartsWith("/"))
            {
                await HandleCommand(client, chatId, text, cancellationToken);
                return;
            }

            Stats.Messages++;
            await client.SendTextMessageAsync(chatId, $"🔊 You said: {text}", cancellationToken: cancellationToken);
        }
        else if (update.CallbackQuery is CallbackQuery query)
        {
            await client.AnswerCallbackQueryAsync(query.Id, cancellationToken: cancellationToken);
            long chatId = query.Message.Chat.Id;
            string data = query.Data;

            if (data == "help")
                await client.SendTextMessageAsync(chatId, "Available commands: /help, /start, /echo, /time, /stats, /about", cancellationToken: cancellationToken);
            else if (data == "time")
            {
                string now = DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss") + " UTC";
                await client.SendTextMessageAsync(chatId, $"🕐 Current time: {now}", cancellationToken: cancellationToken);
            }
            else if (data == "stats")
            {
                string text = $"📊 Statistics:\n  Total messages: {Stats.Messages}\n  Commands used: {Stats.Commands}";
                await client.SendTextMessageAsync(chatId, text, cancellationToken: cancellationToken);
            }
        }
    }

    static async Task HandleCommand(ITelegramBotClient client, long chatId, string text, CancellationToken cancellationToken)
    {
        string[] parts = text.Split(' ', 2);
        string cmd = parts[0].ToLower();
        string arg = parts.Length > 1 ? parts[1] : "";

        switch (cmd)
        {
            case "/start":
                var keyboard = new InlineKeyboardMarkup(new[]
                {
                    new[]
                    {
                        InlineKeyboardButton.WithCallbackData("📖 Help", "help"),
                        InlineKeyboardButton.WithCallbackData("⏰ Time", "time")
                    },
                    new[] { InlineKeyboardButton.WithCallbackData("📊 Stats", "stats") }
                });
                await client.SendTextMessageAsync(chatId,
                    "Hello! I'm an Echo Bot. Send me anything and I'll echo it back. Use /help for commands.",
                    replyMarkup: keyboard, cancellationToken: cancellationToken);
                break;

            case "/help":
                string help = "Available commands:\n/start - Welcome message\n/help - Show this help\n/echo <text> - Echo back your text\n/time - Show current server time\n/stats - Show usage statistics\n/about - About this bot";
                await client.SendTextMessageAsync(chatId, help, cancellationToken: cancellationToken);
                break;

            case "/echo":
                Stats.Messages++;
                if (string.IsNullOrEmpty(arg))
                {
                    await client.SendTextMessageAsync(chatId, "Please provide text after /echo", cancellationToken: cancellationToken);
                    return;
                }
                await client.SendTextMessageAsync(chatId, $"🔊 Echo: {arg}", cancellationToken: cancellationToken);
                break;

            case "/time":
                Stats.Commands++;
                string now = DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss") + " UTC";
                await client.SendTextMessageAsync(chatId, $"🕐 Current time: {now}", cancellationToken: cancellationToken);
                break;

            case "/stats":
                Stats.Commands++;
                string statsText = $"📊 Statistics:\n  Total messages: {Stats.Messages}\n  Commands used: {Stats.Commands}";
                await client.SendTextMessageAsync(chatId, statsText, cancellationToken: cancellationToken);
                break;

            case "/about":
                await client.SendTextMessageAsync(chatId, "🤖 Telegram Echo Bot\nBuilt with Telegram.Bot\nMIT License", cancellationToken: cancellationToken);
                break;

            default:
                await client.SendTextMessageAsync(chatId, $"Unknown command: {cmd}. Use /help", cancellationToken: cancellationToken);
                break;
        }
    }

    static Task HandlePollingErrorAsync(ITelegramBotClient client, Exception exception, CancellationToken cancellationToken)
    {
        Console.WriteLine($"Error: {exception.Message}");
        return Task.CompletedTask;
    }
}

File: bot.rb
# bot.rb
require 'telegram/bot'
require 'time'

TOKEN = ENV['BOT_TOKEN']
unless TOKEN
  puts 'BOT_TOKEN environment variable not set'
  exit 1
end

$stats = { messages: 0, commands: 0 }

Telegram::Bot::Client.run(TOKEN) do |bot|
  bot.listen do |message|
    case message
    when Telegram::Bot::Types::Message
      chat_id = message.chat.id
      text = message.text

      next if text.nil?

      if text.start_with?('/')
        parts = text.split(' ', 2)
        cmd = parts[0].downcase
        arg = parts[1] || ''

        case cmd
        when '/start'
          bot.api.send_message(
            chat_id: chat_id,
            text: "Hello! I'm an Echo Bot. Send me anything and I'll echo it back. Use /help for commands.",
            reply_markup: {
              inline_keyboard: [
                [{ text: '📖 Help', callback_data: 'help' }, { text: '⏰ Time', callback_data: 'time' }],
                [{ text: '📊 Stats', callback_data: 'stats' }]
              ]
            }
          )
        when '/help'
          help_text = "Available commands:\n/start - Welcome message\n/help - Show this help\n/echo <text> - Echo back your text\n/time - Show current server time\n/stats - Show usage statistics\n/about - About this bot"
          bot.api.send_message(chat_id: chat_id, text: help_text)
        when '/echo'
          $stats[:messages] += 1
          if arg.empty?
            bot.api.send_message(chat_id: chat_id, text: "Please provide text after /echo")
          else
            bot.api.send_message(chat_id: chat_id, text: "🔊 Echo: #{arg}")
          end
        when '/time'
          $stats[:commands] += 1
          now = Time.now.utc.strftime('%Y-%m-%d %H:%M:%S UTC')
          bot.api.send_message(chat_id: chat_id, text: "🕐 Current time: #{now}")
        when '/stats'
          $stats[:commands] += 1
          txt = "📊 Statistics:\n  Total messages: #{$stats[:messages]}\n  Commands used: #{$stats[:commands]}"
          bot.api.send_message(chat_id: chat_id, text: txt)
        when '/about'
          bot.api.send_message(chat_id: chat_id, text: "🤖 Telegram Echo Bot\nBuilt with telegram-bot-ruby\nMIT License")
        else
          bot.api.send_message(chat_id: chat_id, text: "Unknown command: #{cmd}. Use /help")
        end
      else
        $stats[:messages] += 1
        bot.api.send_message(chat_id: chat_id, text: "🔊 You said: #{text}")
      end

    when Telegram::Bot::Types::CallbackQuery
      chat_id = message.from.id
      data = message.data
      bot.api.answer_callback_query(callback_query_id: message.id)

      case data
      when 'help'
        bot.api.send_message(chat_id: chat_id, text: "Available commands: /help, /start, /echo, /time, /stats, /about")
      when 'time'
        now = Time.now.utc.strftime('%Y-%m-%d %H:%M:%S UTC')
        bot.api.send_message(chat_id: chat_id, text: "🕐 Current time: #{now}")
      when 'stats'
        txt = "📊 Statistics:\n  Total messages: #{$stats[:messages]}\n  Commands used: #{$stats[:commands]}"
        bot.api.send_message(chat_id: chat_id, text: txt)
      end
    end
  end
end
